### PR TITLE
Parser: Replace dynamic-block regex in `do_blocks`

### PR DESCRIPTION
**Pull request by dmsnell**
## Description

Previously we have been using a simplified parse to grab dynamic
blocks and replace them with their rendered content.

Since #8083 we've had a fast default parser which removes the need
for a simplified parse here.

In this patch we're replacing the existing simplified parser in
`do_blocks` with the new default parser. This will open up new
opportunities for working with nested blocks on the server.

## Resolves or addresses:
 - #6751
 - #7247
 - #8760

## Status
 - [ ] it appears like the old parser relies on the raw content in the blocks but the parser spec doesn't provide this. this means that after this patch we might be wiping out all nested block content on render. this needs to be fixed 😉
 - [x] I'm not sure what the `mobile` files are but they mirror this one. can someone explain what those are? are they auto-generated? do we have to update them?
 - [ ] this doesn't replace usage of the simplified parser in `strip_dynamic_blocks()`. we should probably revisit #6170 to prevent introducing another tear-down-built-up double iteration in those simpler functions. for now, I've left in the regex solution built by @aduth 

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.

### CODE REVIEW

File: lib/blocks.php
Comment (georgestephanis) [inline lib/blocks.php]: I think removing this line is causing the unit test failure in https://travis-ci.org/WordPress/gutenberg/jobs/439435606 as it's changing how some whitespace renders?

The only difference is an extra line break showing up in the output, so not a huge deal, but this may be accounting for it.

**Comment by mcsf** [comment]
> I'm not sure what the mobile files are but they mirror this one. can someone explain what those are? are they auto-generated? do we have to update them?

No need to worry about these. Context: #9883.

> it appears like the old parser relies on the raw content in the blocks but the parser spec doesn't provide this. this means that after this patch we might be wiping out all nested block content on render. this needs to be fixed

What's a viable plan for this?

> this doesn't replace usage of the simplified parser in `strip_dynamic_blocks()`

@dmsnell: Considering this and the tight timeline for 4.1, 4.2 and WordPress 5.0, how do you think this current PR should be prioritized? 

> This will open up new opportunities for working with nested blocks on the server.

I'm all for this, but, with prioritization in mind, we need to identify in which way the present work is a requirement for the MVP of #8352. Also, a note to @mtias and I that we should make sure that said issue, _PHP APIs Overview_, is up-to-date for 5.0.

**Comment by pento** [comment]
d7811ca is kind of janky, but it fixes nested blocks. Instead of dumping the parent block HTML into its `innerHTML`, it creates new `innerBlocks` freeform children, with the HTML fragments interspersed in the right place with the actual child blocks.

**Comment by pento** [comment]
I've manually [added these changes to Core](https://core.trac.wordpress.org/changeset/43752), we can sync up the exact behaviour later.

File: lib/blocks.php
Comment (georgestephanis) [inline lib/blocks.php]: The way these conditionals are laid, blocks can either be Dynamic Blocks, OR they can have innerBlocks.  It doesn't permit Dynamic Blocks to have innerBlocks.  Which seems bad.

**Review by georgestephanis — CHANGES_REQUESTED** [review]
Rather than initially rendering dynamic blocks with no concept of innerBlocks, it would probably be better to render any inner blocks first, then pass that as content to the dynamic block.

This would solve the issues w/r/t Jetpack Contact Forms not having field render in https://github.com/Automattic/jetpack/pull/10256

File: lib/blocks.php
Comment (dmsnell) [reply lib/blocks.php]: this is true, hence the reason this PR exists.

I'd rather not render inner blocks first then pass them up because then that limits what we can do with inner blocks. it should be good enough to give blocks their inner blocks in structural form, let them decide if they want to modify or filter them, then render the document.

this PR is just one step in the process - we have lots to overhaul here and maybe I'm wrong but I thought that @mcsf decided to keep this as-is until post merge and leave dynamic blocks the way they are - as in, yesterday or the day before he decided. let's see what he says

**Comment by gziolo** [comment]
> I've manually added these changes to Core, we can sync up the exact behaviour later.

Let's move it to 4.2 and wait for feedback from Core in such case. We are wrapping up 4.1-RC today and it isn't approved so we really don't have a choice. We can still move it back if only it's ready to go before we do the release.

File: lib/blocks.php
Comment (georgestephanis) [inline lib/blocks.php]: Could we just do something like this, to allow individual plugins to override as necessary for more custom functionality that Core doesn't support w/r/t content, html, inner blocks, etc?

```suggestion
		$block = (array) $block;

		if ( $pre_rendered_content = apply_filters( 'pre_do_block', null, $block, $all_blocks ) ) {
			$rendered_content .= $pre_rendered_content;
			$post = $global_post;
			continue;
		}
```

or potentially even `apply_filters( 'pre_do_block_' . $block['blockName'], null, $block, $all_blocks )` to narrow the filter instead?

Comment (georgestephanis) [reply lib/blocks.php]: This could be taken advantage of via something like this:

https://github.com/Automattic/jetpack/pull/10256/commits/a12356bf7ee36cddb2c51cffb3e51861c09cc180

**Comment by youknowriad** [comment]
@mcsf @dmsnell What's the status of this PR. Is it still meant to be included in 4.2?

**Comment by georgestephanis** [comment]
@youknowriad I believe it may have shifted over to #11082 and #11141 instead?  Unsure.

**Pull request closed without merging**
